python/spark_sklearn/group_apply.py

```diff
@@ -0,0 +1,158 @@
+    returned data frames.
+
+    >>> import pandas as pd
+    >>> df = (spark
+    ...     .createDataFrame([Row(course="dotNET", year=2012, earnings=10000),
+    ...                       Row(course="Java",   year=2012, earnings=20000),
+    ...                       Row(course="dotNET", year=2012, earnings=5000),
+    ...                       Row(course="dotNET", year=2013, earnings=48000),
+    ...                       Row(course="Java",   year=2013, earnings=30000)])
+    ...     .select("course", "year", "earnings"))
+    >>> def yearlyMedian(key, vals):
+    ...     all_years = set(vals['year'])
+    ...     # Note that interpolation is performed, so we need to cast back to long.
+    ...     yearly_median = [(key, year, long(vals['earnings'][vals['year'] == year].median()))
+    ...                      for year in all_years]
+    ...     return pd.DataFrame.from_records(yearly_median)
+    >>> gapply(df.groupBy("course"), yearlyMedian, df.schema).orderBy("earnings").show()
+    +------+----+--------+
+    |course|year|earnings|
+    +------+----+--------+
+    |dotNET|2012|    7500|
+    |  Java|2012|   20000|
+    |  Java|2013|   30000|
+    |dotNET|2013|   48000|
+    +------+----+--------+
+    <BLANKLINE>
+    >>> def twoKeyYearlyMedian(course, year, vals):
+    ...     return pd.DataFrame.from_records([(course, year, long(vals["earnings"].median()))])
+    >>> gapply(df.groupBy("course", "year"), twoKeyYearlyMedian, df.schema, "earnings") \
+            .orderBy("earnings").show()
+    +------+----+--------+
+    |course|year|earnings|
+    +------+----+--------+
+    |dotNET|2012|    7500|
+    |  Java|2012|   20000|
+    |  Java|2013|   30000|
+    |dotNET|2013|   48000|
+    +------+----+--------+
+    <BLANKLINE>
+    """
+    import pandas
+    minPandasVersion = '0.7.1'
+    if LooseVersion(pandas.__version__) < LooseVersion(minPandasVersion):
+        raise ImportError('Pandas installed but version is {}, {} required'
+                          .format(pandas.__version__, minPandasVersion))
+
+    # Do a null aggregation to retrieve the keys first (should be no computation)
+    key_cols = grouped_data.agg({}).columns
+
+    if not cols:
+        # Extract the full column list with the parent df
+        javaDFName = "org$apache$spark$sql$RelationalGroupedDataset$$df"
+        parentDF = java_gateway.get_field(grouped_data._jgd, javaDFName)
+        all_cols = DataFrame(parentDF, None).columns
+        key_cols_set = set(key_cols)
+        cols = [col for col in all_cols if col not in key_cols_set]
+
+    if "*" in cols:
+        raise ValueError("cols expected to contain only singular columns")
+
+    if len(set(cols)) < len(cols):
+        raise ValueError("cols expected not to contain duplicate columns")
+
+    if not isinstance(schema, StructType):
+        raise ValueError("output schema should be a StructType")
+
+    inputAggDF = grouped_data.agg({col: 'collect_list' for col in cols})
```

Comment:
I was not familiar with this syntax. Can you use the explicit `dict` constructor?